### ISSUE TITLE
Replace in nested chains leads to stale tasks

### ISSUE THREAD

**Issue by Agalin** [celery/celery#6325]
# Checklist

- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.
Backend: Redis 5.0.6 (Windows test), 6.0.5 (Linux test).
## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [x] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [x] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- #4631

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
Windows:
software -> celery:4.4.7 (cliffs) kombu:4.6.11 py:3.6.8
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Windows arch:64bit, WindowsPE
            kernel version:10 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis:///

result_backend: 'redis:///'
broker_url: 'redis://localhost:6379//'

Linux:
software -> celery:4.4.7 (cliffs) kombu:4.6.11 py:3.6.8
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Linux arch:64bit, ELF
            kernel version:3.10.0-1062.4.1.el7.x86_64 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis:///

result_backend: 'redis:///'
broker_url: 'redis://localhost:6379//'
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: Tested on 3.6.8.
* **Minimal Celery Version**: Tested on 4.4.7.
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: Tested on 5.0.6 and 6.0.5.
* **Minimal Result Backend Version**: Same as broker.
* **Minimal OS and/or Kernel Version**: Tested on Win10 1809 and RHEL7 kernel 3.10.0-1062.4.1.el7.x86_64
* **Minimal Broker Client Version**: Tested using redis-py 3.5.3
* **Minimal Result Backend Client Version**: Same as broker

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==2.6.1
billiard==3.6.3.0
celery==4.4.7
flower==0.9.5
humanize==2.6.0
importlib-metadata==1.7.0
kombu==4.6.11
prometheus-client==0.8.0
pytz==2020.1
redis==3.5.3
tornado==6.0.4
vine==1.3.0
zipp==3.1.0
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
from celery import Celery, canvas


app = Celery()
app.conf.result_backend = app.conf.broker_url = 'redis://'


@app.task(bind=True)
def demo_replace(self, depth=0, first=True, parent_first=True, depth_limit=2):
    print(f"Current id: {self.request.id}, parent id: {self.request.parent_id}, first: {first}, parent first: {parent_first}")
    if depth < depth_limit:
        raise self.replace(canvas.chain(demo_replace.si(depth+1, True, first, depth_limit), demo_replace.si(depth+1, False, first, depth_limit)))
```

```python
demo_replace.delay()
```

</p>
</details>

# Expected Behavior

All tasks should succeed and end in SUCCEDED state. Should be 4 unique IDs in total for `depth_limit=2`.

# Actual Behavior

Single level replace with a chain works as expected.
Replace in the last branch (where root id is replaced) works as expected.
Replace in other branches ends with task calling `replace` not being replaced (task ends in `PENDING` state), both tasks in chain have new ids.
Log for `limit_depth=2`:
```
Current id: 0f60b4d0-8daf-443c-adcd-05e7f4e5f3aa, parent id: None, first: True, parent first: True
Current id: 7eb33f7b-9975-4d16-ac79-2f3b96caf944, parent id: 0f60b4d0-8daf-443c-adcd-05e7f4e5f3aa, first: True, parent first: True
Current id: 2f857878-db0e-4c3e-bed5-c9b14305cc02, parent id: 7eb33f7b-9975-4d16-ac79-2f3b96caf944, first: True, parent first: True
Current id: 56437eb0-76a0-49db-b764-53ccdcab3564, parent id: 2f857878-db0e-4c3e-bed5-c9b14305cc02, first: False, parent first: True
Current id: 0f60b4d0-8daf-443c-adcd-05e7f4e5f3aa, parent id: 56437eb0-76a0-49db-b764-53ccdcab3564, first: False, parent first: True
Current id: 628e9d08-221a-45c2-9525-823402f18705, parent id: 0f60b4d0-8daf-443c-adcd-05e7f4e5f3aa, first: True, parent first: False
Current id: 0f60b4d0-8daf-443c-adcd-05e7f4e5f3aa, parent id: 628e9d08-221a-45c2-9525-823402f18705, first: False, parent first: False
```
There are 5 unique ids even though there should be only 4 (task with id `56437eb0-76a0-49db-b764-53ccdcab3564` should really have `7eb33f7b-9975-4d16-ac79-2f3b96caf944`).

For higher values of `limit_depth` there are more tasks left in `PENDING` (every one that calls `replace` excluding root task). Didn't have time yet to check if any signal is called for those (although Flower marks them as `STARTED` only so probably at least one expected message is not broadcasted).

**Comment by thedrow** [comment]
Can you please go over the mandatory checklist to ensure this is not already fixed?

**Comment by Agalin** [comment]
Same for current master (5.0.0 RC2 f0dca0d). It has also updated dependendies (notably `kombu` to 5.0.1). BTW is it intentional to no longer support `celery worker -A proj` and require `celery -A proj worker`?

<details>
<summary><b><code>celery report</code> Output for master:</b></summary>
<p>

Windows:
```
software -> celery:5.0.0rc2 (singularity) kombu:5.0.1 py:3.6.8
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Windows arch:64bit, WindowsPE
            kernel version:10 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis:///

result_backend: 'redis:///'
broker_url: 'redis://localhost:6379//'
```
Linux:
```
software -> celery:5.0.0rc2 (singularity) kombu:5.0.1 py:3.6.8
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Linux arch:64bit, ELF
            kernel version:3.10.0-1062.4.1.el7.x86_64 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis:///

result_backend: 'redis:///'
broker_url: 'redis://localhost:6379//'
```


</p>
</details>

<details>
<summary><b>Log for the same testcase using master:</b></summary>
<p>

```
[2020-09-07 10:44:18,880: WARNING/ForkPoolWorker-4] Current id: 52a63147-6200-46b4-83dc-10779d0db3a4, parent id: None, first: True, parent first: True
[2020-09-07 10:44:18,918: WARNING/ForkPoolWorker-4] Current id: 71363dab-43cc-4acd-b87e-c1a46e465d83, parent id: 52a63147-6200-46b4-83dc-10779d0db3a4, first: True, parent first: True
[2020-09-07 10:44:18,922: WARNING/ForkPoolWorker-4] Current id: c203a895-9fb4-438a-9318-4f1f1de57d93, parent id: 71363dab-43cc-4acd-b87e-c1a46e465d83, first: True, parent first: True
[2020-09-07 10:44:18,926: WARNING/ForkPoolWorker-4] Current id: b55cbfad-f5e3-4af6-81d8-fa6ce24f83ae, parent id: c203a895-9fb4-438a-9318-4f1f1de57d93, first: False, parent first: True
[2020-09-07 10:44:18,929: WARNING/ForkPoolWorker-4] Current id: 52a63147-6200-46b4-83dc-10779d0db3a4, parent id: b55cbfad-f5e3-4af6-81d8-fa6ce24f83ae, first: False, parent first: True
[2020-09-07 10:44:18,933: WARNING/ForkPoolWorker-4] Current id: 8e5efb24-0087-4943-9c58-9a03be189586, parent id: 52a63147-6200-46b4-83dc-10779d0db3a4, first: True, parent first: False
[2020-09-07 10:44:18,937: WARNING/ForkPoolWorker-4] Current id: 52a63147-6200-46b4-83dc-10779d0db3a4, parent id: 8e5efb24-0087-4943-9c58-9a03be189586, first: False, parent first: False
```
Same as before - `b55cbfad-f5e3-4af6-81d8-fa6ce24f83ae` should be `71363dab-43cc-4acd-b87e-c1a46e465d83`.

</p>
</details>

Anything more you'd want me to check?

**Comment by Agalin** [comment]
The issue seems to be in `canvas.py:_chain.prepare_steps` method. There is logic for setting previous id for the last task but if I understand celery's logic correctly chain of chains is flattened to a single chain so for example code it's `1 -> (replace 1 by 2 1) -> 2 1 -> (replace 2 by 3 2) -> 3 4 1 (here 2 is not a last task so new id is generated) -> replace 1 by 5 1 -> 3 4 5 1`

**Comment by thedrow** [comment]
> BTW is it intentional to no longer support `celery worker -A proj` and require `celery -A proj worker`?

Yes. It will be documented when GA lands.

**Comment by thedrow** [comment]
> The issue seems to be in `canvas.py:_chain.prepare_steps` method. There is logic for setting previous id for the last task but if I understand celery's logic correctly chain of chains is flattened to a single chain so for example code it's `1 -> (replace 1 by 2 1) -> 2 1 -> (replace 2 by 3 2) -> 3 4 1 (here 2 is not a last task so new id is generated) -> replace 1 by 5 1 -> 3 4 5 1` (instead of 3 2 5 1)

You're right that we flatten chains into one chain but I'm not sure I'm following.
Can you walk me through the root cause step by step (preferably by pointing at the lines you think are causing the issue)?

**Comment by Agalin** [comment]
If my interpretation is correct then it's not a simple issue as the whole loop would be to blame.
Loop starts here: https://github.com/celery/celery/blob/master/celery/canvas.py#L702
And issue would be in this if block: https://github.com/celery/celery/blob/master/celery/canvas.py#L746
Task that is replaced isn't necessarily last in a chain due to flattening so this check evaluates to false and freezes with just root id.

**Comment by Agalin** [comment]
Yeah. It's an issue with replacing tasks in the middle of the chain in general.

```python
from celery import Celery, canvas


app = Celery()
app.conf.result_backend = app.conf.broker_url = 'redis://'

@app.task(bind=True)
def demo_replace2(self, ):
    print(f"Current id: {self.request.id}, parent: {self.request.parent_id}, root: {self.request.root_id}")
    self.replace(canvas.chain(demo_replace3.si(), demo_replace3.si()))

@app.task(bind=True)
def demo_replace3(self):
    print(f"Current id: {self.request.id}, parent: {self.request.parent_id}, root: {self.request.root_id}")

```

```python
(demo_replace2.si() | demo_replace3.si()).delay()
```

Code above leads to the following output:
```
[2020-09-07 15:50:00,166: WARNING/SpawnPoolWorker-1] Current id: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6, parent: None, root: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6
[2020-09-07 15:50:00,269: WARNING/SpawnPoolWorker-2] Current id: e509ab4a-0666-48cc-af08-35e22216fd4d, parent: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6, root: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6
[2020-09-07 15:50:00,366: WARNING/SpawnPoolWorker-4] Current id: 4162ad15-847d-42b4-84da-1d8540764895, parent: e509ab4a-0666-48cc-af08-35e22216fd4d, root: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6
[2020-09-07 15:50:00,465: WARNING/SpawnPoolWorker-5] Current id: 66a7e121-c48f-42fe-a63a-b2b75ea65447, parent: 4162ad15-847d-42b4-84da-1d8540764895, root: 95a419c7-6d7b-4d6a-ab7a-382280fbb3e6
```

`66a7e121-c48f-42fe-a63a-b2b75ea65447` (last task) is correctly the one returned as an AsyncResult and ends in success.
But root task with id `95a419c7-6d7b-4d6a-ab7a-382280fbb3e6` (the one that gets replaced) ends as a stale task (ignored due to being replaced) and this id is not assigned to any of the child tasks.

**Comment by thedrow** [comment]
Can you come up with a fix?
It seems like you have a good grasp of what went wrong.

I'm currently focused on #6266 and this does not block us from releasing so I don't have the capacity to deal with this right now.

**Comment by Agalin** [comment]
I'm not sure if it's easy to fix. That loop seems to ask for a refactor since #4631. It's not a blocker for my team as well as we can mostly work around that thanks to task_postrun signal but I'll try to look at it in my free time this week.

And yeah I agree that 5.0 with python2 code cleanup is a higher priority. This one is basically only an issue when you want to trace tasks AND use dynamic task structure at the same time.

**Comment by thedrow** [comment]
FYI the entire canvas code is going to be re-written and new capabilities will be added.
For example: a new concept called fork is introduced. Fork splits execution into two separate canvases.
This allows you to do the following: 
```python
from celery.canvas import fork, group

c = group(a.sig(), b.sig(), c.sig()) | fork(cb1.sig(), cb2.sig())
```

So instead of one callback in this chord you'll have two which execute in parallel and does not wait for eachother to return the results.

**Comment by restless** [comment]
I'm not sure if it's the exact same problem but definitely related.

I've observed that in case a task is replaced with a chain and one of the chain's tasks fails then the original (replaced) tasks ends stale, stuck in "STARTED" state.

Example:

```
from celery import Celery, chain


app = Celery()
app.conf.result_backend = app.conf.broker_url = 'redis://'


@app.task(bind=True)
def dummy(self, param=None):
    print("DUMMY: ", param)


@app.task(bind=True)
def failing(self, param=None):
    print("FAILING: ", param)
    raise Exception("Failure")


@app.task(bind=True)
def rep(self, param=None):
    print("REPLACING: ", param)
    return self.replace(chain(dummy.si(10), failing.si("1"), dummy.si(11)))


chain(
    dummy.si("1"),
    rep.si("1"),
    dummy.si("2")
).delay()
```

What Flower shows for these is:

```
tasks.failing   1a0fe60e-6644-4688-b6e8-f37229b70589    FAILURE ['1']   {}          2024-05-28 08:23:23.989 2024-05-28 08:23:23.989         celery@5d6233972934
tasks.dummy     37b9df37-98a3-44ec-ae12-4bcd85328e51    SUCCESS (10,)   {}  None    2024-05-28 08:23:23.988 2024-05-28 08:23:23.988 0.00    celery@5d6233972934
tasks.rep       60d81e54-ad45-4ccf-ae3f-ebb44d12c38c    STARTED ['1']   {}          2024-05-28 08:23:23.984 2024-05-28 08:23:23.984         celery@5d6233972934
tasks.dummy     6d065a8d-1754-49a5-8a4f-f0609e21a5cf    SUCCESS ('1',)  {}  None    2024-05-28 08:23:23.979 2024-05-28 08:23:23.980 0.01    celery@5d6233972934
```

The problem here is that the replaced task (`task.rep`) is stuck in `STARTED` and I'd rather expect it to disappear because it got replaced.

When there's no `failing` task in the `rep`:

```
@app.task(bind=True)
def rep(self, param=None):
    print("REPLACING: ", param)
    return self.replace(chain(dummy.si(10), dummy.si(11)))
```

everything is fine. Flower displays:

```
tasks.dummy     269e0870-48bb-458f-9377-b0e0e4c5a184    SUCCESS ['2']   {}  None    2024-05-28 08:34:53.170 2024-05-28 08:34:53.170 0.00    celery@5d6233972934
tasks.dummy     415ace5f-620d-40fb-8a32-62543dd7b023    SUCCESS [11]    {}  None    2024-05-28 08:34:53.165 2024-05-28 08:34:53.166 0.01    celery@5d6233972934
tasks.dummy     0621e0c3-cd49-4c77-a5ae-68ac6fe27797    SUCCESS (10,)   {}  None    2024-05-28 08:34:53.164 2024-05-28 08:34:53.165 0.00    celery@5d6233972934
tasks.dummy     e93c2de3-5d38-4aff-8357-5faef9680814    SUCCESS ('1',)  {}  None    2024-05-28 08:34:53.155 2024-05-28 08:34:53.155 0.01    celery@5d6233972934
```
